File: Algorithms/Sorting_Algorithms/3-Way_Quick_Sort/3-Way_Quick_Sort.cs

using System;

class Sort {
	
	static void swap<T>(ref T lhs, ref T rhs)
	{
		T temp;
		temp = lhs;
		lhs = rhs;
		rhs = temp;
	}

	public static void partition(int[] a, int l, int r, ref int i, ref int j)
	{
		i = l - 1;
		j = r;
		int p = l - 1, q = r;
		int v = a[r];

		while (true) {
			
			while (a[++i] < v)
				;

			while (v < a[--j])
				if (j == l)
					break;

		
			if (i >= j)
				break;

		
			swap(ref a[i], ref a[j]);

			
			if (a[i] == v) {
				p++;
				swap(ref a[p], ref a[i]);
			}

			
			if (a[j] == v) {
				q--;
				swap(ref a[j], ref a[q]);
			}
		}


		swap(ref a[i], ref a[r]);

		j = i - 1;
		for (int k = l; k < p; k++, j--)
			swap(ref a[k], ref a[j]);

		i = i + 1;
		for (int k = r - 1; k > q; k--, i++)
			swap(ref a[i], ref a[k]);
	}


	public static void quicksort(int[] a, int l, int r)
	{
		if (r <= l)
			return;

		int i = 0, j = 0;


		partition(a, l, r, ref i, ref j);

		quicksort(a, l, j);
		quicksort(a, i, r);
	}

	
	public static void printarr(int[] a, int n)
	{
		for (int i = 0; i < n; ++i)
			Console.Write(a[i] + " ");
		Console.Write("\n");
	}

	
	static void Main()
	{
		int[] a = { 4, 9, 4, 4, 1, 9, 4, 4, 9, 4, 4, 1, 4 };
		int size = a.Length;
		
		
		printarr(a, size);
		quicksort(a, 0, size - 1);
		printarr(a, size);
	}
}

File: Algorithms/Sorting_Algorithms/3-Way_Quick_Sort/3-Way_Quick_Sort.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class QuickSort {
	
	private static void swap(Comparable[] arr, int first, int second)
	{	// Swaps the two array elements with indices passed as arguments.
		Comparable temp = arr[first];
		arr[first] = arr[second];
		arr[second] = temp;
	}

	
	private static void quickSort3Way(Comparable[] arr, int low, int high)
	{	/*
			arr: Array to be sorted using the Three Way QuickSort Algorithm
			low: The starting index of a subarray
			high: The ending index of a subarray
		*/
		
		// No partitioning for subarrays of size 1 or below.
		if(high <= low) return;

		// Partitioning item value
		Comparable v = arr[low];
		
		// Three variables lesser, i and greater are used for the corresponding partitions to be made.
		// lesser partition i.e. arr[low .. lesser - 1] denotes the partition with items of value less than the partitioning item v.
		// greater partition i.e arr[greater + 1 .. high] denotes the partition with items of value greater than the partitioning item v.
		// equals partition i.e. arr[lesser .. greater] denotes the partition with items of value equal to the partitioning item v.
		// variable i represents the current element that is to be put into one of the above three partitions.
		int lesser = low, i = low + 1, greater = high;

		while(i <= greater)
		{
			int cmp = arr[i].compareTo(v);
			if(cmp < 0) swap(arr, lesser++, i++);
			else if(cmp > 0) swap(arr, i, greater--);
			else i++;
		}
		quickSort3Way(arr, low, lesser-1);
		quickSort3Way(arr, greater+1, high);
	}
	
	private static void printSortedArray(Comparable[] arr)
	{	// Prints the sorted array as output.
		for(int index=0; index<arr.length; index++)
			System.out.print(arr[index]+" ");
		System.out.println();
	}
	
	public static void main(String[] args) throws IOException
	{
		BufferedReader reader = new BufferedReader(new InputStreamReader(System.in));
		String[] arr = reader.readLine().split(" "); // String[] used in this implementation. Any data type implementing the Comparable interface can be used in it's place.
		quickSort3Way(arr, 0, arr.length-1); // Sort the input array using the Three Way QuickSort Algorithm.
		printSortedArray(arr); // Print the output as the sorted array.
	}
}
